doc/porgball.8.in
.\"----------------------------------------------------------------
.\" porgball.8 - porgball man page
.\"----------------------------------------------------------------
.\" Copyright (C) 2015 David Ricart <@PACKAGE_BUGREPORT@>
.\"----------------------------------------------------------------
.\"
.TH PORGBALL "8" "@RELEASEDATE@" "porg @PACKAGE_VERSION@" "System"
.SH NAME
porgball - Binary package support for porg(8).
.SH SYNOPSIS
\fBporgball\fR [OPTIONS] [-a|<packages>]
.br
\fBporgball\fR -e [OPTIONS] <porgballs>
.SH DESCRIPTION
.PP
By default, \fBporgball\fR may be given one or more
package names as arguments, and for each of those packages it creates
a binary tarball containing all the currently
installed files. For this to work the packages have to be
registered in the porg database.
.br
Leading slashes ('/') are stripped from the paths of the files in the
tarballs, so they can be extracted into user-defined directories.
.PP
With option \fB-e\fR, \fBporgball\fR admits one or more previously
created tarballs ("porgballs") as arguments, and installs them into the
system, optionally logging the installation with porg(8).
.PP
\fBNote\fR: Options cannot be joined up, for instance: '\fB-af\fR' is not correct,
type '\fB-a -f\fR' instead.
.PP
Mandatory arguments to long options are mandatory for short options too.
.SH GENERAL OPTIONS
.TP
\fB-h, --help\fR
Show usage information and exit.
.TP
\fB-v, --verbose\fR
Explain what is being done.
.TP
\fB-V, --version\fR
Print version information and exit.
.TP
\fB-L, --logdir\fR=\fIDIR\fR
Porg log directory. Default is '@LOGDIR@', unless option LOGDIR is set
in \fBporgrc(5)\fR.
.SH PORGBALL CREATION OPTIONS
.TP
\fB-a, --all\fR
Create a porgball for each package logged in the porg database.
.TP
\fB-X, --exact-version\fR
Disable expansion of package version.
.br
By default, porgball automatically expands the package names given as arguments, 
by adding all the versions of the package that are registered in the porg 
database.
.br
The option \fB-X\fR inhibits this expansion, so that package names must match 
the basename and the whole version of a registered package.
.TP
\fB-d, --directory\fR=\fIDIR\fR
Create the porgballs in directory DIR (default is '.').
.TP
\fB-g, --gzip\fR
Compress with gzip (default).
.TP
\fB-b, --bzip2\fR
Compress with bzip2.
.TP
\fB-x, --xz\fR
Compress with xz.
.TP
\fB-#\fR
Set the compression level (speed/quality balance). '#' is a number between 1
(faster compression) and 9 (best compression). Default is 9.
.TP
\fB--fast\fR
An alias for -1.
.TP
\fB--best\fR
An alias for -9.
.TP
\fB-f, --force\fR
Force overwrite of existing output files.
.TP
\fB-t, --test\fR
Test intergity of the porgball after creating it.
.TP
\fB-n, --no-porg-suffix\fR
Do not append '.porg' suffix to the name of the porgballs.
For instance, create 'foo-1.0.tar.gz' instead of 'foo-1.0.porg.tar.gz'.
.SH PORGBALL EXTRACTION OPTIONS
.TP
\fB-e, --extract\fR
Extract (install) the given porgballs, in directory '/' by default.
.TP
\fB-l, --log\fR
Log the file extraction with porg, retrieving the appropiate
package name from the name of the porgball.
.TP
\fB-d, --directory\fR=\fIDIR\fR
Extract the files into directory DIR (as the option -C in tar).
.SH FILES
.PP
\fI@sysconfdir@/porgrc\fR - porg configuration file
.br
\fI@LOGDIR@\fR - default porg log directory
.SH AUTHOR
Written by David Ricart (@PACKAGE_BUGREPORT@)
.SH SEE ALSO
porg(8), porgrc(5)


ChangeLog
Version 0.10 (17 May 2016)
--------------------------

	+ Fixed bug: Make variable names in porgrc file really case 
	  insensitive. Now, "LOGDIR", "logdir" and "LogDiR" are all
	  valid and equivalent (thanks Alain Bureau).


Version 0.9 (16 April 2016)
---------------------------

	General changes:

	+ Removed 'make logme' (no longer required to log the own porg 
	  installation). Experimental. Please report any error.
	+ Removed /mnt, /media and /cdrom from the default EXCLUDE (thanks
	  Tomoaki Nishiyama).
	+ Replace mode_t by int when calling va_arg() so that the code compiles
	  in OSX 10.6 (thanks Tomoaki Nishiyama).
	+ Allow lowercase variable names in porgrc (like 'include=...').

	Changes in grop:

	+ Bugfix: When running grop as root, save config file in directory
	  '/root/.config/grop' (it was saved in the ordinary user home dir).
	+ Do not allow running multiple grop instances at the same time.


Version 0.8 (13 April 2015)
---------------------------

	General changes:

	+ Fixed "double DESTDIR" bug (thanks Rene J.V. Bertin).
	+ Added support for MacOS, and perhaps MacOS X (thanks Masahiro Kasahara
	  and Rene J.V. Bertin).

	Changes in porg:

	+ Better option requirement checking.
	+ When calculating total sizes of packages, sizes of hardlinks to installed
	  files are counted only once (thanks Kardakov Slava).

	Changes in scripts:

	+ porgball: Support logging of tarballs from outside of current working
	  directory (thanks Piotr Karbowski).
	+ porgball: Don't crash when there's no logdir (thanks Piotr Karbowski).
	+ paco2porg: Properly lowercase package names (thanks Kardakov Slava).


Version 0.7 (25 June 2014)
--------------------------

	+ Added missing $(DESTDIR)s to Makefiles.

	Changes in porg:

	+ Check for command line option requirements.
	+ Option -U requires -r.
	+ Added short option -j for --log-missing.
	+ Bugfix: Properly deal with big package sizes (> 4G).

	Changes in grop:

	+ Display a startup info dialog when reading big databases.
	+ Added button "porg suffix" to the porgball creation dialog.
	+ When showing results in the "Find file" dialog, double clicking on a
	  row, or pressing of <Return>, makes the treeview in the main window 
	  scroll to the selected package.
	+ Improved the quality of the grop icon.

	Changes in scripts:

	+ Bugfix: porgball failed to process symbolic links to directories.
	+ Bugfix: paco2porg failed to process missing files.


Version 0.6 (16 June 2014)
--------------------------

	Changes in porg:

	+ Bugfix: porg erroniously complained about command line option
	  incompatibility when some options were doubled (like in '-rr').
	+ Improved the man page and usage output of porg, organizing options by
	  groups. Added a new section 'FILE NAMES WITH SPACES' to the man page. 
	  Reading it is recommended.
	+ Removed path /etc/mtab from the default EXCLUDE.
	+ Search package info in configure.log if config.log is not found.
	+ Ask the user when unlogging a package (option -U), unless option -b
	  is enabled.

	Changes in grop:

	+ New icon.

	Changes in scripts:

	+ New command line options --no-porg-suffix and --logdir for porgball.
	+ Let porgball extract files from ordinary tarballs (without suffix
	  '.porg' in their names).
	+ Honour DESTDIR when installing porg_bash_completion (thanks Veeti
	  Paananen).


Version 0.5 (11 June 2014)
--------------------------

	Changes in porg:

	+ Improved checking of command line options compatibility.
	+ Added paths {/run,/mnt,/media,/cdrom,/etc/mtab} to the default EXCLUDE.

	Changes in grop:

	+ Accept command line options (run 'grop -h' for more info).
	+ Added a "Cancel" button to the porgball creation dialog. Fixed some minor
	  bugs.
	+ Save settings in ~/.config/grop/groprc instead of ~/.grop

	Changes in scripts:

	+ Bugfix: Porgball failed when trying to log extractions with porg.
	+ Bugfix (#1): Paco2porg lost package information (thanks Leandro Nini).


Version 0.4 (26 May 2014)
-------------------------

	+ Bugfix: In some cases, tmp files were not removed after installing a 
	  package.
	+ Bugfix: Some files were not registered when logging the own porg
	  installation with "make logme".
	+ Bugfix: The main window of grop was partially invisible.
	+ Bugfix: Porgball failed to extract porgballs in some cases.
	+ New feature: When installing a package, save the environment variables
	  passed to ./configure, and not only the command line options.
	+ New feature: Expand any environment variable (and not only $HOME),
	  in variable LOGDIR in porgrc.
	+ New feature: Added option '-x|--exact-version' to porg and porgball (see 
	  man page).
	+ Improved the code for searching for package information.
	+ Added porg.spec.
	+ Removed the (buggy) superporg script.


Version 0.3 (19 May 2014)
-------------------------

	+ Readded GPL License and Copyright.
	+ Bugfix: Properly log files with spaces in their names.
	+ Bugfix: EXCLUDE variable in porgrc was not parsed correctly.
	+ Bugfix: Skip sourcing of /etc/bash_completion when installing
	  porg_bash_completion, to avoid errors from foreign software.


Version 0.2 (23 April 2014)
---------------------------

	+ Several bugs fixed thanks to the contribution of Masahiro Kasahara:
	  - paco2porg failed when importing paco logs with no logged files
	  - porg hanged up when logging packages and the installer program calls 
	    jemalloc
	+ Fixed bug that prevented logging of files created by programs that don't
	  terminate normally.
	+ Fixed bug with resolution of relative paths.
	+ Removed option '-g|--log-ignore-errors'.
	+ Handle system calls openat(), linkat(), etc... when installing a package.
	+ Do not handle *truncate() system calls when installing a package.
	+ Removed porg.pc from distribution, as porg is not a library.


Version 0.1 (16 March 2014)
---------------------------

	+ Removed question 3 of the FAQ, which discusses logging of directories.
	+ Removed getopt from the distribution. Use native getopt.

grop/porgball.cc
//=======================================================================
// porgball.cc
//-----------------------------------------------------------------------
// This file is part of the package porg
// Copyright (C) 2015 David Ricart
// For more information visit http://porg.sourceforge.net
//=======================================================================

#include "config.h"
#include "porgball.h"
#include "util.h"
#include "mainwindow.h"
#include "porg/file.h"
#include <gtkmm/table.h>
#include <gtkmm/stock.h>
#include <gtkmm/grid.h>
#include <gtkmm/separator.h>
#include <glibmm/iochannel.h>
#include <glibmm/miscutils.h>	// Glib::get_home_dir()
#include <glibmm/fileutils.h>	// Glib::file_open_tmp()
#include <glibmm/stringutils.h>	// Glib::strerror()
#include <fstream>
#include <sys/wait.h>

using std::string;
using namespace Gtk;

Grop::Porgball::Last Grop::Porgball::s_last = 
	{ Glib::get_home_dir(), PROG_GZIP, 6, false, true };


Grop::Porgball::Porgball(Pkg const& pkg, Window& parent)
:
	Dialog("grop :: porgball", parent, true),
	m_pkg(pkg),
	m_label_progress("", 0.02, 0.5),
	m_label_tarball("", 0, 0),
	m_combo_prog(),
	m_combo_level(),
	m_filechooser_button(FILE_CHOOSER_ACTION_SELECT_FOLDER),
	m_button_test("_Integrity test", true),
	m_button_porg_suffix("_porg suffix", true),
	m_button_close(add_button(Stock::CLOSE, RESPONSE_CLOSE)),
	m_button_cancel(add_button(Stock::CANCEL, RESPONSE_CANCEL)),
	m_button_ok(add_button(Stock::OK, RESPONSE_OK)),
	m_progressbar(),
	m_tmpfile(),
	m_pid(0),
	m_close(false),
	m_children()
{
	set_border_width(8);

	// widgets to be unsensitivized when creating the porgball
	m_children.push_back(&m_combo_prog);
	m_children.push_back(&m_combo_level);
	m_children.push_back(&m_filechooser_button);
	m_children.push_back(&m_button_test);
	m_children.push_back(&m_button_porg_suffix);
	m_children.push_back(m_button_close);

	m_label_progress.set_ellipsize(Pango::ELLIPSIZE_MIDDLE);
	m_label_tarball.set_ellipsize(Pango::ELLIPSIZE_MIDDLE);

	m_combo_prog.append("gzip");
	if (Glib::find_program_in_path("bzip2").size())
		m_combo_prog.append("bzip2");
	if (Glib::find_program_in_path("xz").size())
		m_combo_prog.append("xz");

	m_combo_level.append("1 (faster)");
	for (int i = 2; i < 9; ++i)
		m_combo_level.append(Porg::num2str(i));
	m_combo_level.append("9 (better)");

	Box* box_vexpand = manage(new Box());
	Grid* grid = manage(new Grid());
	grid->set_column_spacing(10);
	grid->set_row_spacing(10);
	int row = 0;
	grid->attach(*(manage(new Label("Name:", 0., 0.5))), 0, row, 1, 1);
	grid->attach(m_label_tarball, 1, row, 3, 1);
	grid->attach(m_button_porg_suffix, 1, ++row, 3, 1);
	grid->attach(*(manage(new Label("Save in folder:", 0, 0.5))), 0, ++row, 1, 1);
	grid->attach(m_filechooser_button, 1, row, 3, 1);
	grid->attach(*(manage(new Label("Compression:", 0., 0.5))), 0, ++row, 1, 1);
	grid->attach(m_combo_prog, 1, row, 1, 1);
	grid->attach(*(manage(new Label("Level:", 0., 0.5))), 2, row, 1, 1);
	grid->attach(m_combo_level, 3, row, 1, 1);
	grid->attach(m_button_test, 1, ++row, 4, 1);
	grid->attach(*box_vexpand, 0, ++row, 4, 1);
	grid->attach(m_label_progress, 0, ++row, 3, 1);
	grid->attach(m_progressbar, 3, row, 1, 1);
	grid->attach(*(manage(new Separator())), 0, ++row, 4, 1);

	m_label_tarball.set_hexpand();
	m_filechooser_button.set_hexpand();
	m_combo_prog.set_hexpand();
	m_combo_level.set_hexpand();
	m_progressbar.set_hexpand();
	box_vexpand->set_vexpand();

	m_progressbar.set_pulse_step(0.007);

	get_content_area()->pack_start(*grid, PACK_EXPAND_WIDGET);

	m_combo_prog.signal_changed().connect(sigc::mem_fun
		(this, &Grop::Porgball::set_tarball_suffix));
	m_button_porg_suffix.signal_clicked().connect(sigc::mem_fun
		(this, &Grop::Porgball::set_tarball_suffix));
	m_button_cancel->signal_clicked().connect(sigc::mem_fun
		(*this, &Grop::Porgball::on_cancel));
	
	get_action_area()->set_layout(BUTTONBOX_EDGE);

	if (::close(Glib::file_open_tmp(m_tmpfile, "grop")) < 0)
		m_tmpfile = "/tmp/grop" + Porg::num2str(getpid());

	m_filechooser_button.set_current_folder(s_last.folder);
	m_combo_prog.set_active(s_last.prog);
	m_combo_level.set_active(s_last.level);
	m_button_test.set_active(s_last.test);
	m_button_porg_suffix.set_active(s_last.porg_suffix);

	show_all();
	m_progressbar.hide();
	m_button_cancel->hide();
}


Grop::Porgball::~Porgball()
{
	if (m_pid)
		kill(m_pid, SIGKILL);

	unlink(m_tmpfile.c_str());

	s_last.folder = m_filechooser_button.get_filename();
	s_last.test = m_button_test.get_active();
	s_last.level = m_combo_level.get_active_row_number();
	s_last.prog = m_combo_prog.get_active_row_number();
	s_last.porg_suffix = m_button_porg_suffix.get_active();
}


void Grop::Porgball::instance(Pkg const& pkg, Window& parent)
{
	Grop::Porgball obj(pkg, parent);
	
	while (!obj.m_close && obj.run() == RESPONSE_OK)
		obj.end_create(obj.create_porgball());
}


void Grop::Porgball::on_cancel()
{
	g_return_if_fail(m_pid > 0);

	if (m_pid)
		kill(m_pid, SIGKILL);

	m_button_cancel->hide();
	m_button_ok->show();
}


bool Grop::Porgball::on_delete_event(GdkEventAny*)
{
	if (m_pid && kill(m_pid, SIGSTOP) == 0) {

		if (run_question_dialog("A process is running. "
			"Do you want to terminate it ?", this)) {
			kill(m_pid, SIGKILL);
			m_pid = 0;
			m_close = true;
		}
		else if (kill(m_pid, SIGCONT) == 0)
			return true;
	}

	return false;
}


void Grop::Porgball::set_children_sensitive(bool setting /* = true */)
{
	for (uint i = 0; i < m_children.size(); ++i)
		m_children[i]->set_sensitive(setting);
}


bool Grop::Porgball::create_porgball()
{
	// check whether we have write permissions on the dest. directory
	string dir = m_filechooser_button.get_filename();
	if (access(dir.c_str(), W_OK) < 0) {
		run_error_dialog(dir + ": " + Glib::strerror(errno), this);
		return false;
	}

    string zipfile = dir + "/" + m_label_tarball.get_text();
	string tarfile = zipfile.substr(0, zipfile.rfind("."));
	
	if (!access(zipfile.c_str(), F_OK)) {
		if (!run_question_dialog("File '" + zipfile + "' already exists.\n"
			"Do you want to overwrite it ?", this))
			return false;
	}

	set_children_sensitive(false);
	m_button_ok->hide();
	m_button_cancel->show();

	// get list of logged files

	std::ofstream ftmp(m_tmpfile.c_str());
	if (!ftmp) {
		run_error_dialog("Error opening temporary file '" 
			+ m_tmpfile + "':" + Glib::strerror(errno), this);
		return false;
	}
		
	m_progressbar.show();
	m_label_progress.set_text("Reading logged files");
	main_iter();

	struct stat s;

	for (Pkg::const_iter f(m_pkg.files().begin()); f != m_pkg.files().end(); ++f) {
		if (!lstat((*f)->name().c_str(), &s))
			ftmp << (*f)->name() << "\n";
	}

	if (!ftmp.tellp()) {
		run_error_dialog("Empty package", this);
		return false;
	}
	
	ftmp.close();
	
	// build tar command and run it

	std::vector<string> argv;
	argv.push_back("tar");
	argv.push_back("--create");
	argv.push_back("--file=" + tarfile);
	argv.push_back("--files-from=" + m_tmpfile);
	argv.push_back("--ignore-failed-read");

	m_label_progress.set_text("Creating " + Glib::path_get_basename(tarfile));
	main_iter();

	if (!spawn(argv)) {
		unlink(tarfile.c_str());
		return false;
	}

	// build compression command and run it

	string prog = m_combo_prog.get_active_text();
	string level = Porg::num2str(m_combo_level.get_active_row_number() + 1);

	argv.clear();
	argv.push_back(prog);
	argv.push_back("-" + level);
	argv.push_back("--force");
	argv.push_back(tarfile);
	
	m_label_progress.set_text("Creating " + Glib::path_get_basename(zipfile));
	main_iter();

	if (!spawn(argv)) {
		unlink(tarfile.c_str());
		unlink(zipfile.c_str());
		return false;
	}

	// if needed, build test command and run it

	if (!m_button_test.get_active())
		return true;
	
	argv.clear();
	argv.push_back(prog);
	argv.push_back("--test");
	argv.push_back(zipfile);

	m_label_progress.set_text("Testing " + Glib::path_get_basename(zipfile));
	main_iter();

	return spawn(argv);
}


void Grop::Porgball::end_create(bool done /* = true */)
{
	m_pid = 0;
	set_children_sensitive();
	m_progressbar.hide();
	m_button_ok->show();
	m_button_cancel->hide();
	m_label_progress.set_markup(
		done ? "<span fgcolor=\"darkgreen\"><b>Done</b></span>" : "");
	main_iter();
}


void Grop::Porgball::set_tarball_suffix()
{
	string name = m_pkg.name();
	if (m_button_porg_suffix.get_active())
		name += ".porg";
	name += ".tar";

	switch (m_combo_prog.get_active_row_number()) {
		case PROG_GZIP:		name += ".gz";	break;
		case PROG_BZIP2:	name += ".bz2"; break;
		case PROG_XZ:		name += ".xz";	break;
		default: g_return_if_reached();
	}

	m_label_tarball.set_text(name);
}


bool Grop::Porgball::spawn(std::vector<string>& argv)
{
	int std_err, status;

	Glib::spawn_async_with_pipes(Glib::get_current_dir(), argv,
		Glib::SPAWN_DO_NOT_REAP_CHILD | Glib::SPAWN_SEARCH_PATH,
		sigc::slot<void>(), &m_pid, NULL, NULL, &std_err);

	Glib::RefPtr<Glib::IOChannel> io = Glib::IOChannel::create_from_fd(std_err);
	io->set_close_on_unref(true);

	while (!m_close && waitpid(m_pid, &status, WNOHANG) != -1) {
		m_progressbar.pulse();
		main_iter();
		g_usleep(2000);	
	}

	if (WIFEXITED(status) && WEXITSTATUS(status) != EXIT_SUCCESS) {
		Glib::ustring aux, err = "Error while running the " + argv[0] + " process";
		if (io->read_to_end(aux) == Glib::IO_STATUS_NORMAL)
			err += ":\n\n" + aux;
		run_error_dialog(err, this);
		return false;
	}

	return !WIFSIGNALED(status);
}

